pkg/debug/checkups/tuf.go

```diff
@@ -54,51 +58,160 @@ func (tc *tufCheckup) Run(ctx context.Context, extraFH io.Writer) error {
-	response, err := client.Get(url.String())
+func (tc *tufCheckup) remoteTufMetadata(client *http.Client, tufUrl *url.URL) (string, int, error) {
+	response, err := client.Get(tufUrl.String())
 	if err != nil {
-		return "", err
+		return "", 0, err
 	}
 	defer response.Body.Close()
 
 	var releaseTargets tufRelease
 	if err := json.NewDecoder(response.Body).Decode(&releaseTargets); err != nil {
-		return "", err
+		return "", 0, err
 	}
 
 	upgradePathKey := fmt.Sprintf("launcher/%s/%s/%s/release.json", runtime.GOOS, tuf.PlatformArch(), tc.k.UpdateChannel())
 	hostTargets, ok := releaseTargets.Signed.Targets[upgradePathKey]
 
 	if !ok {
-		return "", fmt.Errorf("unable to find matching release data for %s", upgradePathKey)
+		return "", releaseTargets.Signed.Version, fmt.Errorf("unable to find matching release data for %s", upgradePathKey)
+	}
+
+	return hostTargets.Custom.Target, releaseTargets.Signed.Version, nil
+}
+
+// localTufMetadata inspects the local target metadata to extract the metadata version.
+// We can compare this version to the one retrieved by `remoteTufMetadata` to see whether
+// this installation of launcher is successfully updating its TUF repository.
+func (tc *tufCheckup) localTufMetadata() (int, error) {
+	targetsMetadataFile := filepath.Join(tuf.LocalTufDirectory(tc.k.RootDirectory()), "targets.json")
+	b, err := os.ReadFile(targetsMetadataFile)
+	if err != nil {
+		return 0, fmt.Errorf("reading file %s: %w", targetsMetadataFile, err)
+	}
+
+	var releaseTargets tufRelease
+	if err := json.Unmarshal(b, &releaseTargets); err != nil {
+		return 0, fmt.Errorf("unmarshalling file %s: %w", targetsMetadataFile, err)
+	}
+
+	return releaseTargets.Signed.Version, nil
+}
+
+// versionsInLauncherLibrary returns all updates available in the launcher update directory.
+func (tc *tufCheckup) versionsInLauncherLibrary() ([]string, error) {
+	updatesDir := tc.k.UpdateDirectory()
+	if updatesDir == "" {
+		updatesDir = tuf.DefaultLibraryDirectory(tc.k.RootDirectory())
+	}
+
+	launcherVersionMatchPattern := filepath.Join(updatesDir, "launcher", "*")
+	launcherMatches, err := filepath.Glob(launcherVersionMatchPattern)
+	if err != nil {
+		return nil, fmt.Errorf("globbing for launcher matches at %s: %w", launcherVersionMatchPattern, err)
+	}
+
+	return launcherMatches, nil
+}
+
+// versionsInOsquerydLibrary returns all updates available in the osqueryd update directory.
+func (tc *tufCheckup) versionsInOsquerydLibrary() ([]string, error) {
+	updatesDir := tc.k.UpdateDirectory()
+	if updatesDir == "" {
+		updatesDir = tuf.DefaultLibraryDirectory(tc.k.RootDirectory())
+	}
+
+	osquerydVersionMatchPattern := filepath.Join(updatesDir, "osqueryd", "*")
+	osquerydMatches, err := filepath.Glob(osquerydVersionMatchPattern)
+	if err != nil {
+		return nil, fmt.Errorf("globbing for osqueryd matches at %s: %w", osquerydVersionMatchPattern, err)
+	}
+
+	return osquerydMatches, nil
+}
```

Comment:
I'd probably merge these and take the match string.

Reply:
The match string will be long/annoying to read in that case -- e.g. for me it'd be `"/var/kolide-k2/k2device-preprod.kolide.com/updates/launcher/1.1.2,/var/kolide-k2/k2device-preprod.kolide.com/updates/launcher/1.1.2-13-g2789a54,/var/kolide-k2/k2device-preprod.kolide.com/updates/launcher/1.1.2-15-gaa6f573"`. (With the array approach, whatever I use to read the JSON file will split it up onto separate lines and make it much more readable.) If you'd rather a string, I think I should iterate through and call `filepath.Base` on each so we end up with `"1.1.2,1.1.2-13-g2789a54,1.1.2-15-gaa6f573"`. Which do you prefer?

Reply:
(I'd been kind of wanting to do `filepath.Base` anyway so I'm happy with that plus a join!)

Reply:
I don't feel strongly. I noticed the duplicated functions that look like they differ only by a the `"launcher"` or `"osqueryd"` strings.

Reply:
_oh_

Reply:
Sorry, I keep missing which lines a comment applies to on the diff in GitHub -- updated to combine

Reply:
GitHub can be _something_